interfaces/builtin/microstack_support.go

```diff
@@ -0,0 +1,245 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+/*
+ * Microstack is a full OpenStack in a single snap package.
+ * Virtual machines are spawned as QEMU processes with libvirt acting as a management
+ * daemon (including for activities such as applying AppArmor profiles).
+ * Networking is provided largely via OpenVSwitch and Neutron with dnsmasq acting
+ * as an auxiliary daemon. tun/tap kernel module is used for creating virtual interfaces.
+ * Virtual machines rely on KVM for virtualization acceleration and on vhost
+ * framework in the kernel (vhost_net, vhost_scsi, vhost_vsock).
+ */
+
+const microStackSupportSummary = `allows operating as the MicroStack service`
+
+const microStackSupportBaseDeclarationPlugs = `
+  microstack-support:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const microStackSupportBaseDeclarationSlots = `
+  microstack-support:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const microStackSupportConnectedPlugAppArmor = `
+
+# Used by QEMU to work with the kernel-side virtio implementations.
+/dev/vhost-net rw,
+/dev/vhost-scsi rw,
+/dev/vhost-vsock rw,
+# Used by QEMU to work with VFIO (https://www.kernel.org/doc/Documentation/vfio.txt).
+# For vfio hotplug on systems without static vfio (LP: #1775777)
+# VFIO userspace driver interface.
+/dev/vfio/vfio rw,
+# Access to VFIO group character devices such as /dev/vfio/<group> where <group> is the group number.
+/dev/vfio/* rw,
+# Used by Nova for mounting images via qemu-nbd.
+/dev/nbd* rw,
+
+# Allow creating dm-* devices, /dev/<vg-name> directories, /dev/mapper directory and symlinks under it.
+# Allow issuing ioctls to the Device Mapper for LVM tools via /dev/mapper/control.
+/dev/** rw,
```

Comment:
This is quite a dangerous rule. Can we somehow play with openstack or LVM so that the virtual group directories are created with names following a certain pattern?

Reply:
Volume groups are created using `vgcreate` which uses has a static template for naming a directory that would contain LV symlinks for a particular VG: `/dev/<vgname>`

https://github.com/lvmteam/lvm2/blob/v2_03_07/lib/activate/fs.c#L42-L45
https://github.com/lvmteam/lvm2/blob/v2_03_07/lib/activate/fs.c#L275-L279

LVM volume group names are configurable in OpenStack and users may configure as many instances of this volume driver as they want (we have an experimental loop-based backend for testing that has a specific name but there are no restrictions as to how users can configure their own one).

https://docs.openstack.org/cinder/latest/configuration/block-storage/drivers/lvm-volume-driver.html#id1
volume_group = <vg-name>

The usage scenario of the LVM driver for MicroStack is to provide an ability for a user to configure storage as they want it and provide a VG on top of it for use by MicroStack. So restricting VG names would be limiting the way users might want to use or name things.

To summarize: the issue is mostly that LVM has a static location that it wants to be able to access and there are no means to control it.

Reply:
Plan A would be to modify the LVM tools to use a different pattern for the names (like, change that format string to be `%s/vg_%s`); I guess the tools are part of the snap, so we could patch them.

Or, in alternative:

> The usage scenario of the LVM driver for MicroStack is to provide an ability for a user to configure storage as they want it and provide a VG on top of it for use by MicroStack. So restricting VG names would be limiting the way users might want to use or name things.

Yes, but isn't this a compromise we could make? I mean, in the apparmor profile we could allow those names which are used by the most popular MicroStack services, and also allow a certain pattern for user chosen names (for instance, that volume groups start with `vg`). As long as we find a way to document this, do you still see an issue?

Reply:
> do you still see an issue

For this interface, yes, since the MicroStack snap also gets `block_devices` interface access which means `rw` [access](https://github.com/snapcore/snapd/blob/2.51.2/interfaces/builtin/block_devices.go#L60-L65) to all block devices. As such, it is free to overwrite any signatures or contents on the backing block devices, including the data of any volumes in any of the volume groups. This makes the volume group name restriction artificial since MicroStack is the only user of the `microstack-support` interface.

>  I guess the tools are part of the snap, so we could patch them.

I went through the pain of upgrading the snap to core20 and newer versions of different parts (as a part of changing the base series and OpenStack version) and a significant amount of time was spend on rebasing/reworking all the patches we carry already - it would be preferable to avoid even more patches since we do not have capacity to carry this forward all the time.

It would be better to focus on allowing libvirt running as a part of MicroStack to apply AppArmor policies to QEMU processes it spawns but this does not work at the moment due to a different issue https://forum.snapcraft.io/t/call-chains-loose-ld-library-path-and-thereby-child-programs-fail/24382. We never had this enabled in MicroStack so it can be an incremental step.

> allow those names which are used by the most popular MicroStack services

That's the thing - OpenStack does not configure the LVM backend itself. It is completely up to the deployer which VG name gets chosen and there may be preexisting VGs that people might want to use.

In summary: because MicroStack has enough privileges to circumvent the vg name restriction you are talking about, I would not create artificial restrictions which could later result in bug reports. LXD does not have such restrictions while it [can work with LVM](https://github.com/lxc/lxd/blob/lxd-4.16/lxd/storage/drivers/driver_lvm.go) and in this case we are dealing with the hypervisor software so it would make sense for it to be able to use arbitrary VG names.